src/test_runner.ts

```diff
@@ -0,0 +1,93 @@
+import * as vscode from "vscode";
+import { spawn } from "child_process";
+import { get_build_directory } from "./extension_helpers";
+import { test_details } from "./test_details";
+
+async function run_test(
+  test: vscode.TestItem,
+  signal: AbortSignal,
+  log: vscode.OutputChannel
+): Promise<{ stdout: string; stderr: string; code: number | null }> {
+  return new Promise((resolve, reject) => {
+    const test_detail = test_details.get(test)!;
+    const command = test_detail.command[0];
+    const args = test_detail.command.slice(1);
+    const process = spawn(command, args, {
+      signal,
+      cwd: get_build_directory(),
+    });
+
+    let stdout =
+      command + " " + args.map((val) => JSON.stringify(val)).join(" ") + "\n";
+    let stderr = "";
+    if (process.pid) {
+      process.stdout.on("data", (data) => (stdout += data.toString()));
+      process.stderr.on("data", (data) => (stderr += data.toString()));
+      process.on("close", (code) => resolve({ stdout, stderr, code }));
+      process.on("error", (err) => reject(err));
+    } else {
+      reject({ message: `Failed to run test ${command} ${args.join(" ")}` });
+    }
+  });
+}
+
+/**
+ * The run handler executes each requested test and reports the output
+ * or errors to the VS Code Testing API via the TestRun interface. Tests
+ * may be cancelled via the CancellationToken.
+ * @param test_controller VS Code test controller
+ * @param log VS Code extension output channel
+ * @param shouldDebug Should tests run in debug
+ * @param request VS Code test run request
+ * @param token VS Code cancellation token
+ */
+export async function run_tests(
+  test_controller: vscode.TestController,
+  log: vscode.OutputChannel,
+  shouldDebug: boolean,
+  request: vscode.TestRunRequest,
+  token: vscode.CancellationToken
+) {
+  const abort_controller = new AbortController();
+  const { signal } = abort_controller;
```

Comment:
What do the braces do on this line?

Reply:
It's shorthand for `const signal = abort_controller.signal`

It's handy for getting multiple values from an object at one time. It can also be used in a function's arguments

```ts
function foo ({a, c}) { ... }
const config = {a: 1, b: 2, c: 3}
foo(config)
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment